File: 7.py
from copy import deepcopy
import itertools
import re
from collections import namedtuple

bag_pair = namedtuple('bag_pair', ['name','nr'])
d = {}
rules = {}
allbags = set()
for s in open('7.in','r').readlines():
    pattern = "^(\S+ \S+){1}"
    match = re.search(pattern, s)
    parent = match[0]
    pattern = "(\d \S+ \S+ bags?)"
    match = re.findall(pattern, s)
    children = []
    for m in match:
        l = m.split() # "4 drab silver bags" -> ['4','drab','silver','bags']
        child = bag_pair(l[1]+' '+l[2],int(l[0])) # ("drab silver", 4)
        children.append(child)
    rules[parent] = children

def one():
    wcsg = set() # who contains shiny gold?
    repeat = True
    to_find = {"shiny gold"}
    while repeat:
        repeat = False
        for r in rules:
            for c in rules[r]:
                if c.name in to_find and r not in to_find:
                    wcsg.add(r)
                    to_find.add(r)
                    repeat = True
    return len(wcsg)

# Recursive way to solve first part, 50x slower :/
# def eventually_contains(entry, bag):
#     retval = False
#     for r in rules[entry]:
#         if r.name==bag: return True
#         retval = eventually_contains(r.name, bag)
#         if retval: return True
#     return retval

def count_bags_in(bag):
    bags = 0
    for r in rules[bag]:
        bags += r.nr*(1 + count_bags_in(r.name))
    return bags

def two():
    return count_bags_in("shiny gold")
    

if __name__ == "__main__":
    print(f"{one() = }") # 103
    print(f"{two() = }") # 1469

File: 7.cpp
#include "aoc.hpp"

using namespace std;

const string inputname = "7.in";
using bag_t = string;
using tree_bag_t = unordered_map<bag_t, vector<pair<bag_t,int>>>;
using sol_t = int;
tree_bag_t tree_bag;
sol_t one()
{
    std::set<bag_t> wcsg{};
    std::set<bag_t> to_find{"shiny gold"};
    bool repeat = true;
    while (repeat)
    {
        repeat = false;
        for (auto & r: tree_bag)
        {
            for (auto & c: tree_bag[r.first])
            {
                if ((to_find.find(c.first)!=to_find.end()) &&
                    (to_find.find(r.first)==to_find.end()))
                {
                    wcsg.insert(r.first);
                    to_find.insert(r.first);
                    repeat = true;
                }
            }
        }
    }
    return wcsg.size();
}

int count_bags_in(bag_t bag)
{
    auto children = tree_bag[bag];
    int sum = 0;
    for (auto & c: children)
    {
        sum += c.second*(1+count_bags_in(c.first));
    }
    return sum;
}

sol_t two()
{
    return count_bags_in("shiny gold");
}


int main(int argc, char **kwargs)
{
    ifstream infile(inputname);
    bag_t s, key, bag_name;
    regex rc("(\\d)|(\\S+ \\S+)(?:(?!bags contain)\\w)+"); // it should not work, but it sort of does  [muted white] [bags contain] [3] [muted tomato] [5] [light black] [4] [pale black] [5] [shiny gold]
    while( getline( infile, s ) ) // fo r each line read from the file
    {
        std::regex_token_iterator<std::string::iterator> rend;
        std::regex_token_iterator<std::string::iterator> b ( s.begin(), s.end(), rc);
        bag_t parent = *b++;
        //skip "bags contain", god forbid I managed to remove with the regex
        b++;
        vector<pair<bag_t,int>> child_bags{};

        if ((*b).str()!="no other")
        {
            while (b!=rend)
            {
                pair<bag_t,int> bp;
                bp.second = stoi((*b++).str());
                bp.first = (*b++).str();
                child_bags.push_back(bp);
            }
        }
        tree_bag[parent] = child_bags;
    }
    std::cout << "one = " << one() << '\n'; // 
    std::cout << "two = " << two() << '\n'; // 

    return 0;
}

